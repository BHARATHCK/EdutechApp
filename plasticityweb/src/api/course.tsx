import { Course } from "../types";
import redaxios from "redaxios";

export async function createCourse(params: {
  courseName: string;
  courseType: string;
  notes: string;
  isPublished: boolean;
  author: string;
  video: string;
}): Promise<Course> {
  const response = await redaxios.post(
    `${process.env.REACT_APP_BACKEND_URL}/createcourse`,
    { params },
    { withCredentials: true },
  );
  return response.data;
}

export async function getAllCourses(): Promise<Course[]> {
  const response = await redaxios.get(`${process.env.REACT_APP_BACKEND_URL}/courses`, {
    withCredentials: true,
  });
  return response.data;
}

export async function getCourse(params: { courseId: number }): Promise<Course> {
  const response = await redaxios.post(
    `${process.env.REACT_APP_BACKEND_URL}/course`,
    { params },
    { withCredentials: true },
  );
  return response.data;
}

export async function deleteCourse(params: { courseId: number }): Promise<Course> {
  const response = await redaxios.post(
    `${process.env.REACT_APP_BACKEND_URL}/deleteCourse`,
    { params },
    { withCredentials: true },
  );

  return response.data;
}
