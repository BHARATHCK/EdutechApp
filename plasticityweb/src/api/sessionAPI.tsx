import { User } from "../types";
import redaxios from "redaxios";

export async function login(loginData: any) {
  const user = await fetch("http://localhost:3000/login", loginData);
  console.log("RESPONSE login : ", user);
  return user.json();
}

export async function Login(params: { email: string; password: string }): Promise<User> {
  const response = await redaxios.post(
    "http://localhost:3000/login",
    { params },
    { withCredentials: true },
  );

  console.log("RESPONSE Login : ", response);

  return response.data;
}

export async function logout() {
  const response = await redaxios.delete("/api/sessions");

  return response.data.data;
}
