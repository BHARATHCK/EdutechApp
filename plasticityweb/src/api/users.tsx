import redaxios from "redaxios";
import { User } from "../types";

export async function getCurrentUser(): Promise<User> {
  const response = await redaxios.get("http://localhost:3000/me", { withCredentials: true });
  return response.data;
}

export async function signUp(params: {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
}): Promise<User> {
  const response = await redaxios.post(
    "http://localhost:3000/signup",
    { params },
    { withCredentials: true },
  );
  return response.data;
}
