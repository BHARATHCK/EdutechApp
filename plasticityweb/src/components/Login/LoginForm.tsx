import { Button, FormControl, FormErrorMessage, FormLabel, Input } from "@chakra-ui/react";
import { Field, Form, Formik } from "formik";
import * as React from "react";
import "./LoginForm.styles.scss";
import * as Yup from "yup";

const LoginForm = () => {
  const SignupSchema = Yup.object().shape({
    firstName: Yup.string().min(2, "Too Short!").max(50, "Too Long!").required("Required"),
    lastName: Yup.string().min(2, "Too Short!").max(50, "Too Long!").required("Required"),
    email: Yup.string().email("Invalid email").required("Required"),
    password: Yup.string()
      .required("No password provided.")
      .min(8, "Password is too short - should be 8 chars minimum.")
      .matches(/[a-zA-Z]/, "Password can only contain Latin letters."),
    passwordConfirmation: Yup.string()
      .oneOf([Yup.ref("password")], "Passwords must match")
      .required("Required"),
  });

  async function postData(url = "", data = {}) {
    // Default options are marked with *
    const response = await fetch(url, {
      method: "POST",
      mode: "cors",
      cache: "no-cache",
      credentials: "same-origin",
      headers: {
        "Content-Type": "application/json",
      },
      redirect: "follow",
      referrerPolicy: "no-referrer",
      body: JSON.stringify(data),
    });
    return response.json(); // parses JSON response into native JavaScript objects
  }

  return (
    <div className="loginForm">
      <Formik
        initialValues={{
          firstName: "",
          lastName: "",
          email: "",
          password: "",
          passwordConfirmation: "",
        }}
        validationSchema={SignupSchema}
        onSubmit={(values, { setSubmitting, setFieldError }) => {
          console.log("Sending request : ", values);
          postData("http://localhost:3000/users", values).then((data) => {
            console.log(data);
            setSubmitting(false);
          });
        }}
      >
        {(props) => (
          <Form>
            <Field name="firstName">
              {({ field, form }: any) => (
                <FormControl isInvalid={form.errors.firstName && form.touched.firstName}>
                  <FormLabel htmlFor="firstName">First name</FormLabel>
                  <Input size="md" {...field} id="firstName" placeholder="First Name" />
                  <FormErrorMessage>{form.errors.firstName}</FormErrorMessage>
                </FormControl>
              )}
            </Field>
            <Field name="lastName">
              {({ field, form }: any) => (
                <FormControl isInvalid={form.errors.lastName && form.touched.lastName}>
                  <FormLabel htmlFor="lastName">Last name</FormLabel>
                  <Input {...field} id="lastName" placeholder="Last name" />
                  <FormErrorMessage>{form.errors.lastName}</FormErrorMessage>
                </FormControl>
              )}
            </Field>
            <Field name="email">
              {({ field, form }: any) => (
                <FormControl isInvalid={form.errors.email && form.touched.email}>
                  <FormLabel htmlFor="email">Email</FormLabel>
                  <Input {...field} id="email" placeholder="Email" />
                  <FormErrorMessage>{form.errors.email}</FormErrorMessage>
                </FormControl>
              )}
            </Field>
            <Field name="password">
              {({ field, form }: any) => (
                <FormControl isInvalid={form.errors.password && form.touched.password}>
                  <FormLabel htmlFor="password">Password</FormLabel>
                  <Input {...field} id="password" type="password" placeholder="Password" />
                  <FormErrorMessage>{form.errors.password}</FormErrorMessage>
                </FormControl>
              )}
            </Field>
            <Field name="passwordConfirmation">
              {({ field, form }: any) => (
                <FormControl
                  isInvalid={form.errors.passwordConfirmation && form.touched.passwordConfirmation}
                >
                  <FormLabel htmlFor="passwordConfirmation">Password</FormLabel>
                  <Input
                    {...field}
                    id="passwordConfirmation"
                    type="password"
                    placeholder="Re-Type Password"
                  />
                  <FormErrorMessage>{form.errors.passwordConfirmation}</FormErrorMessage>
                </FormControl>
              )}
            </Field>
            <Button mt={4} colorScheme="teal" isLoading={props.isSubmitting} type="submit">
              Submit
            </Button>
          </Form>
        )}
      </Formik>
    </div>
  );
};

export default LoginForm;
