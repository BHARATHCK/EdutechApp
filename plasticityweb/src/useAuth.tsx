import React, { createContext, ReactNode, useContext, useEffect, useMemo, useState } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import * as sessionsApi from "./api/sessionAPI";
import * as usersApi from "./api/users";
import { User } from "./types";

interface AuthContextType {
  user?: User;
  loading: boolean;
  error?: any;
  login: (email: string, password: string) => any;
  signUp: (
    email: string,
    firstName: string,
    lastName: string,
    password: string,
    role: string,
  ) => any;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType>({} as AuthContextType);

// Export the provider to wrap the entire app with it.
export function AuthProvider({ children }: { children: ReactNode }): JSX.Element {
  const [user, setUser] = useState<User>();
  const [error, setError] = useState<any>();
  const [loading, setLoading] = useState<boolean>(false);
  const [loadingInitial, setLoadingInitial] = useState<boolean>(true);

  const navigate = useNavigate();
  const location = useLocation();

  // reset the error state.
  useEffect(() => {
    if (error) setError(null);
  }, [location.pathname]);

  // On initial load check if there is an active session in backend for current user based on cookies
  useEffect(() => {
    usersApi
      .getCurrentUser()
      .then((responseUser: User) => {
        setUser(responseUser);
      })
      .catch((_error: any) => {})
      .finally(() => setLoadingInitial(false));
  }, []);

  // Login function to verify user and set user state on success and redirect to home.
  function login(email: string, password: string) {
    setLoading(true);
    const response = new Promise((resolve, reject) => {
      sessionsApi
        .Login({ email, password })
        .then((user: User) => {
          setUser(user);
          navigate("/");
          resolve(user);
        })
        .catch((error: any) => {
          setError(error);
          reject(error);
        })
        .finally(() => setLoading(false));
    });

    return response;
  }

  // SignUp function to create user, On success update user state and redirect user
  function signUp(
    email: string,
    firstName: string,
    lastName: string,
    password: string,
    role: string,
  ) {
    setLoading(true);
    const newResp = new Promise((resolve, reject) => {
      usersApi
        .signUp({ email, firstName, lastName, password, role })
        .then((user: User) => {
          setUser(user);
          navigate("/");
          resolve(user);
        })
        .catch((error: any) => {
          setError(error);
          reject(error);
        })
        .finally(() => {
          setLoading(false);
        });
    });

    return newResp;
  }
  // Logout
  function logout() {
    sessionsApi.logout().then(() => setUser(undefined));
  }

  // memo the user auth related data.
  const memoedValue = useMemo(
    () => ({
      user,
      loading,
      error,
      login,
      signUp,
      logout,
    }),
    [user, loading, error],
  );

  // If user is available already the memo will be used on re-renders
  return (
    <AuthContext.Provider value={memoedValue}>{!loadingInitial && children}</AuthContext.Provider>
  );
}

// Export useAuth for accessing the context
export default function useAuth() {
  return useContext(AuthContext);
}
