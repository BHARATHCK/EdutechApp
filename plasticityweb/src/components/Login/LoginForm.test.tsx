import { fireEvent, render, waitFor } from "@testing-library/react";
import React from "react";
import { BrowserRouter } from "react-router-dom";
import "@testing-library/jest-dom/extend-expect";
import LoginForm from "./LoginForm";
import useAuth, { AuthProvider } from "../../useAuth";
import { Login } from "../../api/sessionAPI";
import * as userUtils from "../../api/sessionAPI";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { create } from "react-test-renderer";
import { ChakraProvider, theme } from "@chakra-ui/react";

const getAuthFunctions = () => {
  const { login } = useAuth();
  return login;
};

describe("Login Form", () => {
  it("renders correctly", () => {
    let renderer = create(
      <MemoryRouter initialEntries={["/login"]}>
        <Routes>
          <Route path="login" element={<LoginForm />}></Route>
        </Routes>
      </MemoryRouter>,
    );

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});

it("Email and Password should validation on blur", async () => {
  const { getByLabelText, getByTestId } = render(
    <BrowserRouter>
      <LoginForm />
    </BrowserRouter>,
  );
  const email = getByLabelText("Email");
  const password = getByLabelText("Password");

  fireEvent.blur(email);
  fireEvent.blur(password);

  await waitFor(() => {
    expect(getByTestId("emailError")).not.toBe(null);
    expect(getByTestId("emailError")).toHaveTextContent("Required");
    expect(getByTestId("passwordError")).not.toBe(null);
    expect(getByTestId("passwordError")).toHaveTextContent("No password provided.");
  });
});

it("Show validation errors on Submit", async () => {
  const { getByText, getByTestId } = render(
    <BrowserRouter>
      <LoginForm />
    </BrowserRouter>,
  );

  const button = getByTestId("loginButton");
  fireEvent.click(button);

  await waitFor(() => {
    expect(getByTestId("emailError")).toHaveTextContent("Required");
    expect(getByTestId("passwordError")).toHaveTextContent("No password provided.");
  });
});

it("Should remove validation errors on entering values", async () => {
  const { getByText, getByTestId, queryByTestId, getByLabelText } = render(
    <BrowserRouter>
      <LoginForm />
    </BrowserRouter>,
  );

  const button = getByTestId("loginButton");
  fireEvent.click(button);

  const email = getByLabelText("Email");
  const password = getByLabelText("Password");

  fireEvent.change(email, { target: { value: "mockemail@email.com" } });
  fireEvent.change(password, { target: { value: "TESTPPassword" } });

  await waitFor(() => {
    expect(queryByTestId("emailError")).toBe(null);
    expect(queryByTestId("passwordError")).toBe(null);
  });
});

describe("Login Form with context", () => {
  it("routed to login page with context", () => {
    let renderer = create(
      <BrowserRouter>
        <AuthProvider>
          <MemoryRouter initialEntries={["/login"]}>
            <Routes>
              <Route path="login" element={<LoginForm />}></Route>
            </Routes>
          </MemoryRouter>
        </AuthProvider>
      </BrowserRouter>,
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});

it("Test Authentication validation errors", async () => {
  const { getByText, getByTestId, queryByTestId, getByLabelText } = render(
    <BrowserRouter>
      <LoginForm />
    </BrowserRouter>,
  );

  const email = getByLabelText("Email");
  const password = getByLabelText("Password");

  fireEvent.change(email, { target: { value: "mockemail@email.com" } });
  fireEvent.change(password, { target: { value: "TESTPPassword" } });

  const button = getByText("Log In");
  fireEvent.click(button);

  const mockuserData = {
    firstName: "bharath",
    lastName: "chandra",
    email: "bharathchandra541@gmail.com",
    role: "student",
    isActive: true,
    id: 37,
    password: "",
  };
  const mockGetLoginInfo = jest.spyOn(userUtils, "Login");
  mockGetLoginInfo.mockResolvedValue(mockuserData);

  await waitFor(() => {
    //expect(Login).toHaveBeenCalledTimes(1);
    expect(queryByTestId("emailError")).toBe(null);
    expect(queryByTestId("passwordError")).toBe(null);
  });
});
