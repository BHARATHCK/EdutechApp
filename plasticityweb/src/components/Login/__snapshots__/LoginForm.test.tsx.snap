// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form renders correctly 1`] = `
<div
  className="css-gmuwbf"
>
  <div
    className="loginForm css-1s0xmce"
  >
    <form
      action="#"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="chakra-form-control css-0"
        role="group"
      >
        <label
          className="chakra-form__label css-uvho5g"
          htmlFor="email"
          id="field-1-label"
        >
          Email
        </label>
        <input
          className="chakra-input css-0"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          readOnly={false}
          required={false}
          value=""
        />
      </div>
      <div
        className="chakra-form-control css-0"
        role="group"
      >
        <label
          className="chakra-form__label css-uvho5g"
          htmlFor="password"
          id="field-1-label"
        >
          Password
        </label>
        <input
          className="chakra-input css-0"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Password"
          readOnly={false}
          required={false}
          type="password"
          value=""
        />
      </div>
      <div
        className="css-dvxtzn"
      >
        <button
          className="chakra-button css-gz9hhv"
          data-testid="loginButton"
          type="submit"
        >
          Log In
        </button>
        <button
          className="chakra-button css-1tlxye4"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          Join Plasticity Instead
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Login Form with context routed to login page with context 1`] = `null`;
